topdown/bits_test.go

```diff
@@ -0,0 +1,94 @@
+// Copyright 2019 The OPA Authors.  All rights reserved.
+// Use of this source code is governed by an Apache2
+// license that can be found in the LICENSE file.
+
+package topdown
+
+import "testing"
+
+func TestBuiltinBitsOr(t *testing.T) {
+	tests := []struct {
+		note     string
+		rules    []string
+		expected interface{}
+	}{
+		{"basic bitwise-or", []string{`p[x] { x := bits.or(7, 9) }`}, `[15]`},
+		{"or with zero is value", []string{`p[x] { x := bits.or(50, 0) }`}, `[50]`},
+	}
+
+	for _, tc := range tests {
+		runTopDownTestCase(t, map[string]interface{}{}, tc.note, tc.rules, tc.expected)
+	}
+}
+
+func TestBuiltinBitsAnd(t *testing.T) {
+	tests := []struct {
+		note     string
+		rules    []string
+		expected interface{}
+	}{
+		{"basic bitwise-and", []string{`p[x] { x := bits.and(7, 9) }`}, `[1]`},
+		{"and with zero is and", []string{`p[x] { x := bits.and(50, 0) }`}, `[0]`},
+	}
+
+	for _, tc := range tests {
+		runTopDownTestCase(t, map[string]interface{}{}, tc.note, tc.rules, tc.expected)
+	}
+}
+
+func TestBuiltinBitsNegate(t *testing.T) {
+	tests := []struct {
+		note     string
+		rules    []string
+		expected interface{}
+	}{
+		{"basic bitwise-negate", []string{`p[x] { x := bits.negate(42) }`}, `[-43]`},
+	}
+
+	for _, tc := range tests {
+		runTopDownTestCase(t, map[string]interface{}{}, tc.note, tc.rules, tc.expected)
+	}
+}
+
+func TestBuiltinBitsXOr(t *testing.T) {
+	tests := []struct {
+		note     string
+		rules    []string
+		expected interface{}
+	}{
+		{"basic bitwise-xor", []string{`p[x] { x := bits.xor(42, 3) }`}, `[41]`},
+		{"xor same is 0", []string{`p[x] { x := bits.xor(42, 42) }`}, `[0]`},
+	}
+
+	for _, tc := range tests {
+		runTopDownTestCase(t, map[string]interface{}{}, tc.note, tc.rules, tc.expected)
+	}
+}
+
+func TestBuiltinBitsShiftLeft(t *testing.T) {
+	tests := []struct {
+		note     string
+		rules    []string
+		expected interface{}
+	}{
+		{"basic shift-left", []string{`p[x] { x := bits.lsh(1, 3) }`}, `[8]`},
```

Comment:
It might be worthwhile to add a test case that would cause something smaller than big.int's to overflow to ensure that we don't regress anywhere in the future by transitioning through something smaller (like golang 32 or 64 bit int) and lose bits.

Reply:
@patrick-east, I'm at a bit of an impasse here and could use assistance, if you have any recommendations.

I tested `MaxInt32` shifted left oncinpasse (which would cause issues if we were just using int32s) and everything worked okay.

I went to try `MaxInt64` using a test like the following

``` go
{
	"shift of max int64 doesn't overflow",
	[]string{fmt.Sprintf(`p = x { x := bits.lsh(%d, 1) }`, math.MaxInt64)},
	`18446744073709551614`,
},
```

However, since the ast's using `json.Number` under the hood (seems to just be a type alias for `string`), `math.MaxInt64` gets stored as `"9.223372037e+18'`. Then later, when I try to parse it using `NumberToInt` which uses the following under-the hood

``` go
new(big.Int).SetString(string(n), 10)
```
`big.Int` can't parse it.

Any ideas? Am I missing something?

Reply:
Hmmm, yea so I guess other places we always use `big.Float`'s which has a `Parse` helper https://golang.org/pkg/math/big/#Float.Parse so it "just works". What we might need to do is something like:

* Get `ast.Number` operand (ie `json.Number`)
* Get `big.Float` by parsing the number
* Get a `big.Int` from the `big.Float` via https://golang.org/pkg/math/big/#Float.Int and validate that the returned accuracy is `Exact` (otherwise we didn't get a valid int to start with).

So maybe if we update `NumberToInt` to be more like:
```
func NumberToInt(n ast.Number) (*big.Int, error) {
	f := NumberToFloat(n)
	r, accuracy := f.Int(nil)
	if accuracy != big.Exact {
		return nil, fmt.Errorf("illegal value")
	}
	return r, nil
}
```

Does that work?

Reply:
@patrick-east that almost works.

The test I posted is outputting the following
```
            Got: 18446744074000000000
            Exp: 18446744073709551614
```

I believe that something odd is happening with big integers in general here. I totally could be wrong but I think go's JSON library is actually _lossy_ in that it's converting big integers to exponent-notation.

For example, here's another test (just using modulus to try to recover a digit that I believe is being lost)

```
		{
			"big numbers JSON?",
			[]string{fmt.Sprintf(`p = x { x := [%d %% 10, %d] }`, math.MaxInt64, math.MaxInt64)},
			fmt.Sprintf("[%d, %d]", math.MaxInt64%10, math.MaxInt64),
		},
```

It outputs a failure of

```
            Got: [0 9.223372037e+18]
            Exp:
            [7 9223372036854775807]
```

Reply:
Hmm yea for sure we do lossy things when we transition between them in the builtin helpers, for example: https://github.com/open-policy-agent/opa/blob/db922a74e3a4e6f61d739e567c68f69815d2ef9d/topdown/builtins/builtins.go#L154-L157 uses the `.String()` method that defaults to limit the precision to `10` in https://golang.org/pkg/math/big/#Float.String when it calls https://golang.org/pkg/math/big/#Float.Text (similar for the int's)

I think the golang JSON library just keeps numbers as strings so it might be ok if we fix how we convert them and just give longer strings.

This reminded me that we have https://github.com/open-policy-agent/opa/issues/501 which has been open for a long time.

I guess for now maybe we just add a note to document this behavior with large numbers and add test cases for the truncated results (they'll need to be updated when #501 is addressed). I don't want to block this going in for this kind of systematic problem with large numbers in opa.

Reply:
@patrick-east, that sounds good. I added that test.